### ISSUE TITLE
runtime/cgo: should not set -Werror

### ISSUE THREAD

**Issue by williamh** [golang/go#63903]
### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.21.3 linux/amd64
</pre>

### Does this issue reproduce with the latest release?

This was reported by a user a while back for go 1.19, but looking at the source for 1.21.3 as indicated in the [gentoo bug](https://bugs.gentoo.org/887285), the issue still exists.

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env

```
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/william/.cache/go-build'
GOENV='/home/william/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/william/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/william/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/lib/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='local'
GOTOOLDIR='/usr/lib/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.3'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='x86_64-pc-linux-gnu-gcc'
CXX='x86_64-pc-linux-gnu-g++'
CGO_ENABLED='1'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build1224117578=/tmp/go-build -gno-record-gcc-switches'
```
</pre></details>

### What did you do?

The only thing I can provide is the comment in the gentoo bug linked above.
The distro policy is that we shouldn't be setting -Werror on any compilations.

### What did you expect to see?

cgo seems to force -Werror into all compilations of c code.

### What did you see instead?

This should not happen according to our distro policy linked in the bug because it causes compilations to break much more often than necessary.

I don't know if you are willing to make this change upstream or not, but if you aren't, can you please advise me as the maintainer on Gentoo with regard to making this change?

Thanks for your time,

William

**Comment by cherrymui** [comment]
> cgo seems to force -Werror into all compilations of c code.

The `-Werror` in the runtime/cgo package only affects the C code in the runtime/cgo package. It is not adding `-Werror` to user C code. Are you seeing `-Werror` causing failures to building the Go standard library itself?

**Comment by williamh** [comment]
@cherrymui I have not, but here is another [bug](https://bugs.gentoo.org/887229) that has been marked as a duplicate and seems to show that -Werror is added to user code.

**Comment by cherrymui** [comment]
The `build.log.xz` from that bug seems to indicate that all `-Werror` are applied to the `runtime/cgo` package itself.

**Comment by williamh** [comment]
Ok, that makes sense.
If I set CGO_CFLAGS to some value, would that replace the CFLAGS that are being applied to this code? I have a pending proposal to do that.

**Comment by cherrymui** [comment]
> If I set CGO_CFLAGS to some value, would that replace the CFLAGS that are being applied to this code?

Yes and no. CGO_CFLAGS will be added to the C compilations, but not replacing the existing flags. I don't think we document/guarantee the order of flags that are passed. So if you set `CGO_CFLAGS=-Wno-error`, it may or may not cancel that `-Werror`.

Just to add, from the build.log.xz, the `-Werror` doesn't seem to cause any compilation failures.

**Comment by williamh** [comment]
I realize it isn't causing any compilation errors, but I'm getting flagged by our distro's qa just because it is there. Their [documentation](https://devmanual.gentoo.org/ebuild-writing/common-mistakes/index.html#-werror-compiler-flag-not-removed) suggests dropping it for releases and keeping it for development.
Can you give me a way to disable the adding of -Werror, maybe an environment variable or a command line switch that the make scripts can use?

**Comment by cherrymui** [comment]
I don't think we currently have such a way. One way that I can think of is to have a wrapper script for the C compiler that scans the flags and drops `-Werror` before invoking the actual C compiler, and set `CC` to the wrapper script.

Are you proposing we introduce a way to do that?

**Comment by williamh** [comment]
I'm not sure that would be a good way to do it since we already set CC, so I'm going to bring one of our toolchain guys into this bug.

@thesamesam What do you think?

**Comment by cherrymui** [comment]
Policy-wise, I understand that some people may not agree with using `-Werror`. But I also don't really understand why it is a problem if one wants their own code to be compiled with `-Werror` (not other's code), or in general how their own code should be compiled. Thanks.

**Comment by thesamesam** [comment]
The issue is it makes building Go (or something using Go, given this came up when building e.g. Podman) a pain when testing new compilers which may introduce new warnings.

**Comment by williamh** [comment]
@cherrymui Do you have any thoughts on this?

**Comment by cherrymui** [comment]
Personally I don't have strong opinion about `-Werror`. But I think people frequently build Go with new versions of GCC and clang, so if a newer version of GCC or clang introduces new warnings, the failure is reported in a timely base. If you test a new version of GCC or clang and see such failures, you are also welcome to report. Or are you testing more customized C toolchains?

**Comment by torarnv** [comment]
With `CGO_ENABLED=1` and trying to build a darwin executable on Linux I get:

```
clang: error: argument unused during compilation: '-arch arm64' [-Werror,-Wunused-command-line-argument]
```

And setting `CGO_CFLAGS="-Wno-error"` does not override this, as the `-Werror` is added by go after that.

**Comment by thesamesam** [comment]
> If you test a new version of GCC or clang and see such failures, you are also welcome to report. Or are you testing more customized C toolchains?

Sorry, I thought I'd answered this. Yeah, if it's simply a new _default_ warning with GCC or Clang, we should report it and get it fixed in Go if appropriate, but if we're e.g. customising the toolchain to enable our own stuff by default, that becomes more problematic. It's also then a pain for keeping older Go versions working.

**Comment by ianlancetaylor** [comment]
Not to miss the obvious, but I would expect that you could avoid the clang error with `CGO_CFLAGS=-Wno-unused-command-line-argument` or `CGO_CFLAGS=-w`.

That said, I think that we should always put `CGO_CFLAGS` last on the command line.

**Comment by Laky-64** [comment]
I'm getting the same issue when trying to use MSVC, I can't use MSVC with the -Werror flag since isn't supported for MSVC, so the build fails, any update regarding this issue?
